src/steamship/agents/functional/output_parser.py

```diff
@@ -0,0 +1,89 @@
+import json
+import re
+from typing import Dict, List, Optional
+
+from steamship import Block, MimeTypes, Steamship
+from steamship.agents.schema import Action, AgentContext, FinishAction, OutputParser, Tool
+from steamship.data.tags.tag_constants import RoleTag
+
+
+# TODO(dougreid): extract shared bits from this and the ReACT output parser into a utility?
+class FunctionsBasedOutputParser(OutputParser):
+
+    tools_lookup_dict: Optional[Dict[str, Tool]] = None
+
+    def __init__(self, **kwargs):
+        tools_lookup_dict = {tool.name: tool for tool in kwargs.pop("tools", None)}
```

Comment:
```suggestion
        tools_lookup_dict = {tool.name: tool for tool in kwargs.pop("tools", [])}
```

defaulting to None will cause an exception to be raised since it's not iterable. Unsure if that's intentional. If it isn't we can default to an empty array.

Reply:
I think I copy/pasted this from the other bit, so I'll need to update that too then.